worker/src/task.rs

```diff
@@ -0,0 +1,600 @@
+use async_shutdown::Shutdown;
+use futures::Future;
+use once_cell::sync::OnceCell;
+use parking_lot::Mutex;
+use std::collections::BTreeMap;
+use std::{pin::Pin, sync::Arc};
+use tokio::task::JoinHandle;
+use tracing::{error, info, warn};
+
+#[derive(Clone, PartialEq)]
+enum TaskTrackerState {
+    Running,
+    ShutdownStarted,
+    ShutdownComplete,
+}
+
+#[derive(Clone)]
+struct Task {
+    task_id: u64,
+    tag: Option<String>,
+    shutdown: Shutdown,
+}
+
+/// The type of error that is returned from tasks in this module
+#[allow(dead_code)]
+pub type Error = tokio::sync::oneshot::error::TryRecvError;
+
+pub struct TaskHandle<T> {
+    id: u64,
+    shutdown: Shutdown,
+    join_handle: JoinHandle<Result<T, Error>>,
+    detached: bool,
+}
+
+impl<T> TaskHandle<T> {
+    pub async fn stop(&self) {
+        self.shutdown.shutdown();
+        self.shutdown.wait_shutdown_complete().await;
+    }
+    pub fn stop_no_wait(&self) {
+        self.shutdown.shutdown();
+    }
+    pub fn detach(&mut self) {
+        self.detached = true;
+    }
+    pub fn id(&self) -> u64 {
+        self.id
+    }
+}
+
+impl<T> Future for TaskHandle<T> {
+    type Output = Result<T, Error>;
+
+    fn poll(
+        mut self: Pin<&mut Self>,
+        cx: &mut std::task::Context<'_>,
+    ) -> std::task::Poll<Self::Output> {
+        use futures::future::FutureExt;
+        // If the task panics just propagate it up
+        self.join_handle.poll_unpin(cx).map(Result::unwrap)
+    }
+}
+
+impl<T> Drop for TaskHandle<T> {
+    fn drop(self: &mut Self) {
+        if !self.detached {
+            self.stop_no_wait();
+        }
+    }
+}
+
+/// TaskTracker is a wrapper around tokio for asynchronous tasks and futures.
+/// This class adds the following functionalities which are not present in tokio:
+/// * Tokio does not provide a way to selectively wait for certain tasks to
+///   finish executing while cancelling others during system shutdown in an easy
+///   way without keeping track of all join handles
+/// * Tokio does not provide an api to run some cleanup code when cancelling a task
+/// * TaskTracker provides an easy api to cancel a bunch of spawned tasks with a user
+///   provided tag which would otherwise be difficult to do by keeping track of
+///   tokio returned join handles
+///
+/// It is recommended to use task tracker as a global singleton while it is
+/// certainly possible to have multiple instances of it. It is possible to clone
+/// and share it across multiple tasks. Each clone uses the same internal state.
+/// When we are ready to shudown, just call shutdown_and_wait() and wait for the
+/// returned future to complete.
+/// WARNING: Dropping a task tracker without invoking `shutdown_and_wait()` first
+/// will result in a panic
+///
+/// # Examples
```

Comment:
You could probably convert those to actual document tests (ex like we do in https://github.com/MystenLabs/narwhal/blob/c0eafd241c368b7cba73bec9d350a037068c755e/primary/src/block_waiter.rs#L114)

Reply:
done